src/main/java/address/sync/CloudSimulator.java

```diff
@@ -1,129 +1,670 @@
-                person.setCity(java.util.UUID.randomUUID().toString());
-                person.setStreet(java.util.UUID.randomUUID().toString());
-                person.setPostalCode(java.util.UUID.randomUUID().toString());
+    @Override
+    public RawCloudResponse editTag(String addressBookName, String oldTagName, CloudTag updatedTag) {
+        if (shouldSimulateNetworkFailure()) return getNetworkFailedResponse();
+        if (shouldSimulateSlowResponse()) delayRandomAmount();
+
+        int noOfRequestsRequired = 1;
+        if (!isWithinQuota(noOfRequestsRequired)) {
+            rateLimitStatus.setQuotaRemaining(0);
+            return getEmptyResponse(HttpURLConnection.HTTP_FORBIDDEN);
+        }
+        rateLimitStatus.useQuota(noOfRequestsRequired);
+        try {
+            CloudAddressBook fileData = readCloudAddressBookFromFile(addressBookName);
+            CloudTag returnedTag = updateTagDetails(fileData.getAllTags(), oldTagName, updatedTag);
+            writeCloudAddressBookToFile(addressBookName, fileData);
+
+            modifyCloudTagBasedOnChance(returnedTag);
+            return new RawCloudResponse(HttpURLConnection.HTTP_OK, convertToInputStream(returnedTag),
+                                        convertToInputStream(getStandardHeaders()));
+        } catch (NoSuchElementException e) {
+            return getEmptyResponse(HttpURLConnection.HTTP_BAD_REQUEST);
+        } catch (JAXBException e) {
+            return getEmptyResponse(HttpURLConnection.HTTP_INTERNAL_ERROR);
+        }
+    }
+
+    /**
+     * Deletes a tag uniquely identified by its name, if quota is available
+     *
+     * Consumes 1 API usage
+     *
+     * @param addressBookName
+     * @param tagName
+     * @return
+     */
+    @Override
+    public RawCloudResponse deleteTag(String addressBookName, String tagName) {
+        if (shouldSimulateNetworkFailure()) return getNetworkFailedResponse();
+        if (shouldSimulateSlowResponse()) delayRandomAmount();
+
+        int noOfRequestsRequired = 1;
+        if (!isWithinQuota(noOfRequestsRequired)) {
+            rateLimitStatus.setQuotaRemaining(0);
+            return getEmptyResponse(HttpURLConnection.HTTP_FORBIDDEN);
+        }
+        rateLimitStatus.useQuota(noOfRequestsRequired);
+        try {
+            CloudAddressBook fileData = readCloudAddressBookFromFile(addressBookName);
+            deleteTagFromData(fileData.getAllTags(), tagName);
+            writeCloudAddressBookToFile(addressBookName, fileData);
+            return getEmptyResponse(HttpURLConnection.HTTP_NO_CONTENT);
+        } catch (NoSuchElementException e) {
+            return getEmptyResponse(HttpURLConnection.HTTP_BAD_REQUEST);
+        } catch (JAXBException e) {
+            return getEmptyResponse(HttpURLConnection.HTTP_INTERNAL_ERROR);
+        }
+    }
+
+    @Override
+    public RawCloudResponse initializeAddressBook(String addressBookName) {
```

Comment:
This is an API not in GitHub. We require this since there is no way to add an addressbook without having to manually add a cloud file. Should this also cost API usage?